### PR TITLE
[libc++] <experimental/simd> Fix conversion constructor

### LINKED ISSUE #74327: [libc++] Experimental simd tests make assumptions about floating point conversions
I encountered this issue while working on https://github.com/llvm/llvm-project/pull/68753. Basically, some simd tests start failing when optimizations are enabled because they make invalid assumptions about floating-point conversions.

Reproducer (from `libcxx/test/std/experimental/simd/simd.class/simd_ctor_conversion.pass.cpp`):

```
#include <bit>
#include <cassert>
#include <cstddef>
#include <type_traits>
#include <utility>
#include <iostream>

struct simd {
  long double __data __attribute__((__vector_size__(std::bit_ceil((sizeof(long double) * 2)))));

  template <class Array>
  simd(Array const& array) noexcept {
    for (size_t __i = 0; __i != 2; ++__i) {
      __data[__i] = array[__i];
    }
  }
};

int main(int, char**) {
  double as_double __attribute__((__vector_size__(std::bit_ceil((sizeof(double) * 2)))));
  as_double[0] = static_cast<double>(0);
  as_double[1] = static_cast<double>(1);

  simd as_long_double = as_double;

  std::cout << "data[0] as double = " << as_double[0] << std::endl;
  std::cout << "data[1] as double = " << as_double[1] << std::endl;

  std::cout << "data[0] as long double = " << as_long_double.__data[0] << std::endl;
  std::cout << "data[1] as long double = " << as_long_double.__data[1] << std::endl;

  std::cout << "0.0l = " << 0.0l << std::endl;
  std::cout << "1.0l = " << 1.0l << std::endl;

  assert(as_long_double.__data[0] == 0.0l);
  assert(as_long_double.__data[1] == 1.0l);

  return 0;
}
```

Godbolt: https://godbolt.org/z/8vaKaMb4s

In order to reproduce this in the test suite, you can run `libcxx/test/std/experimental/simd/simd.class/simd_ctor_conversion.pass.cpp` with `-O3` by adding `// ADDITIONAL_COMPILE_FLAGS: -O3` to the test and it should reproduce. You should do this inside the Docker image environment to make sure it reproduces, see `libcxx/utils/ci/run-buildbot-container`.

### PR DESCRIPTION
**Pull request by joy2myself**
Fixes #74327

### CODE REVIEW

**Comment by ldionne** [comment]
This fails on MinGW now: https://buildkite.com/llvm-project/libcxx-ci/builds/32079#018c4260-fae8-40b0-b082-9a4b9edf5723.

So yeah I think we should fix the test itself that uses equality to compare `long double`s obtained via two different routes.

File: libcxx/test/std/experimental/simd/simd.class/simd_ctor_conversion.pass.cpp [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 // UNSUPPORTED: c++03, c++11, c++14
 // XFAIL: target=powerpc{{.*}}le-unknown-linux-gnu
+// ADDITIONAL_COMPILE_FLAGS: -O3
```
Comment (ldionne) [inline libcxx/test/std/experimental/simd/simd.class/simd_ctor_conversion.pass.cpp:11]: We should remove this before landing this patch. We don't want to hardcode the optimization setting in this test. When I land my PR we'll have a CI configuration that tests everything with optimizations enabled so it'll cover that.

**Comment by joy2myself** [comment]
@ldionne If I understand correctly, you want me to compare `long double`s within a error tolerance. I pushed a the code about it. But in my local test, this does not seem to solve the problem under the `-O3` option. The `long double` does not seem to have an error, but is not written correctly.

**Comment by ldionne** [comment]
> @ldionne If I understand correctly, you want me to compare `long double`s within a error tolerance. I pushed a the code about it. But in my local test, this does not seem to solve the problem under the `-O3` option. The `long double` does not seem to have an error, but is not written correctly.

Perhaps you need to adjust the tolerance? Or can you re-organize the tests such that you don't rely on comparing values that went through conversions?

**Comment by joy2myself** [comment]
https://godbolt.org/z/84vahGGaT

It looks like this issue cannot be resolved by adjusting tolerance. Moreover, this problem not only occurs under the `-O3` option but also persists under the `-O2` option. I attempted to compile the same code using `clang-15` and `clang-16` with the same options, as shown in the link above. It can be observed that the expected value of `data[1] as long double`, which should be `1`, is actually output as `7.81847e-4942`. The difference in datas is beyond any tolerance range. Additionally, the code compiled with `clang-15` has no issues, so I am inclined to believe that this is a bug introduced by `clang-16` in the `-O2` optimization.

**Pull request closed without merging**